client/layout/masterbar/checkout.tsx

```diff
@@ -73,9 +78,36 @@ const CheckoutMasterbar: FunctionComponent< Props > = ( {
 			console.error( 'Error getting query string in close button' );
 		}
 
+		if ( closeUrl.startsWith( '/' ) ) {
+			page( closeUrl );
+			return;
+		}
 		window.location.href = closeUrl;
 	}, [ siteSlug, checkoutBackUrl, previousPath, dispatch ] );
 
+	const cartKey = useCartKey();
+	const { responseCart, replaceProductsInCart } = useShoppingCart( cartKey );
+	const [ isModalVisible, setIsModalVisible ] = useState( false );
+	const clickClose = () => {
+		if ( responseCart.products.length > 0 ) {
+			setIsModalVisible( true );
+			return;
+		}
+		leaveCheckout();
+	};
+
+	const modalTitleText = String(
+		translate( 'You are about to leave checkout with items in your cart' )
+	);
+	const modalBodyText = String(
+		translate( 'You can leave the items in the cart or clear your cart.' )
+	);
+	const modalPrimaryText = String( translate( 'Continue' ) );
+	const modalSecondaryText = String( translate( 'Clear cart' ) );
```

Comment:
If the types in `CheckoutModalProps` were not `string`s but `React.ReactNode`s, you wouldn't need to cast them with `String()`.